File: 2024/7.py
import operator
from collections.abc import Callable

import util


def dfs(test: int, nums: list[int], operators: list[Callable[[int, int], int]]) -> bool:
    stack = [(nums[0], 1)]
    while stack:
        res, i = stack.pop()
        if res > test:
            continue

        if i == len(nums):
            if res == test:
                return True
            continue

        for op in operators:
            stack.append((op(res, nums[i]), i + 1))

    return False


def calibrate(lines: list[str], operators: list[Callable[[int, int], int]]) -> int:
    ans = 0

    for line in lines:
        test, rest = line.split(":")

        if dfs(test := int(test), list(map(int, rest.split())), operators):
            ans += test

    return ans


def part1(input: str) -> int:
    return calibrate(input.splitlines(), [operator.add, operator.mul])


def part2(input: str) -> int:
    return calibrate(
        input.splitlines(),
        [operator.add, operator.mul, lambda x, y: int(str(x) + str(y))],
    )


if __name__ == "__main__":
    util.run(7, part1, part2)

File: 2025/7.ts
import { run } from "./util";

async function part1(input: string): Promise<number> {
  const grid = input.split("\n").map((line) => line.split(""));

  let splits = 0;
  for (let r = 1; r < grid.length; r++) {
    const row = grid[r];
    for (let c = 0; c < row.length; c++) {
      if (grid[r - 1][c] === "S" || grid[r - 1][c] === "|") {
        switch (grid[r][c]) {
          case ".":
            grid[r][c] = "|";
            break;
          case "^":
            if (grid[r][c - 1] == ".") {
              grid[r][c - 1] = "|";
            }
            if (grid[r][c + 1] == ".") {
              grid[r][c + 1] = "|";
            }
            splits += 1;
            break;
        }
      }
    }
  }
  return splits;
}

async function part2(input: string): Promise<number> {
  const grid = input.split("\n").map((line) => line.split(""));

  const memo = new Map<string, number>();

  function dfs(r: number, c: number): number {
    if (r === grid.length) {
      return 1;
    }

    const key = `${r},${c}`;
    let val = memo.get(key);
    if (val !== undefined) return val;

    let res;
    const curr = grid[r][c];
    switch (curr) {
      case ".":
        res = dfs(r + 1, c);
        break;
      case "^":
        res = dfs(r + 1, c - 1) + dfs(r + 1, c + 1);
        break;
      default:
        res = 0;
    }

    memo.set(key, res);
    return res;
  }

  return dfs(1, grid[0].indexOf("S"));
}

run(7, part1, part2);
